### PR TITLE
Linux: bump minimum kernel version to 4.18

### PR DESCRIPTION
**Pull request by robn**
### Motivation and Context

I started cutting things out in #16434. Everyone loved it, and wanted more. I'm nothing if not a crowd pleaser!

### Description

Just like last time, but this time the minimum is 4.18, as the base kernel for RHEL8.

Since last time, more stuff removed. Following comments there, I've left `DECLARE_EVENT_CLASS`, and the baseline checks.

### How Has This Been Tested?

Compile checked on kernel.org kernels:

* 4.19.319
* 5.4.281
* 5.10.223
* 5.15.164
* 6.1.102
* 6.2.16
* 6.3.13
* 6.4.16
* 6.5.13
* 6.6.43
* 6.7.12
* 6.8.12
* 6.9.12
* 6.10.3
* 6.11.0

And on following distros:

* 5.4.0-84-generic (Ubuntu 18.04)
* 5.13.0-30-generic (Ubuntu 20.04)
* 5.15.0-25-generic (Ubuntu 22.04)
* 6.8.0-31-generic (Ubuntu 24.04)
* 4.18.0-348.7.1.el8_5.x86_64 (CentOS 8.4.2105)
* 4.18.0-553.8.1.el8_10.x86_64 (Rocky 8.6)
* 5.14.0-427.28.1.el9_4.x86_64 (Rocky 9)
* 6.10.5-200.fc40.x86_64 (Fedora 40)

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by tonyhutter** [comment]
Did you intend to remove `config/intlmacosx.m4`?

File: config/kernel-config-defined.m4 [hunk 1]
```diff
@@ -4,21 +4,6 @@ dnl # detected at configure time and cause a build failure.  Otherwise
...
-				return !license_is_gpl_compatible(
-				    "$ZFS_META_LICENSE");
-			])
-		], [
-			AC_DEFINE([ZFS_IS_GPL_COMPATIBLE], [1],
-			    [Define to 1 if GPL-only symbols can be used])
-		], [
-		])
-	])
-
```
Comment (tonyhutter) [inline config/kernel-config-defined.m4:21]: This seems unrelated to >=4.18 kernels.

**Comment by tonyhutter** [comment]
Other than my comments above, I don't see any issues.  Thank you for all this hard work!

File: config/kernel-config-defined.m4 (hunk 1 above)
Comment (robn) [reply config/kernel-config-defined.m4:21]: Its general cleanups - nothing checks the `ZFS_IS_GPL_COMPATIBLE`, and I can't see that anything ever did.

**Comment by robn** [comment]
> Did you intend to remove `config/intlmacosx.m4`?

Yes. Again, I can't see that anything ever used the defines (`HAVE_CFPREFERENCESCOPYAPPVALUE`, `HAVE_CFLOCALECOPYCURRENT`, `HAVE_CFLOCALECOPYPREFERREDLANGUAGES`) or the substitution (`INTL_MACOSX_LIBS`), and there's nothing in the history to suggest why it was added.

**Comment by tonyhutter** [comment]
> Its general cleanups - nothing checks the ZFS_IS_GPL_COMPATIBLE, and I can't see that anything ever did.

>  Again, I can't see that anything ever used the defines (HAVE_CFPREFERENCESCOPYAPPVALUE, HAVE_CFLOCALECOPYCURRENT, HAVE_CFLOCALECOPYPREFERREDLANGUAGES) or the substitution (INTL_MACOSX_LIBS), and there's nothing in the history to suggest why it was added.

Ah I see now.  Yea, looks like dead code, let's keep it removed.

**Review by tonyhutter — APPROVED** [review]

**Review by mcmilk — APPROVED** [review]
Needs a rebase and some mods within these files:
./module/os/linux/zfs/zpl_file.c
./config/kernel-vfs-direct_IO.m4
./config/kernel.m4
./config/kernel-vfs-iov_iter.m4

LGTM :+1:

**Comment by robn** [comment]
Yep, was on it last night but I got tired.

Last push rebased. Compile checks have been redone (kernel versions updated in original post), and results checked again.

I've got a couple more rounds of testing in progress just to be totally sure, but I'm pretty confident this is good.

**Comment by behlendorf** [comment]
Thanks for the quick rebase of this Rob.  I'll review this today so we can get it in.

**Comment by amotin** [comment]
As I have mentioned in the Direct I/O PR, it seems `config/kernel-get-user-pages.m4` added there could be completely removed.

**Comment by mcmilk** [comment]
Debian 11 seems to have problems:
* https://github.com/mcmilk/zfs/actions/runs/10888703273/job/30213756532
* https://github.com/mcmilk/zfs/actions/runs/10888703273/job/30224840461

Since the two runs fail the same way, there seems sth. bad to happen :(

**Review by behlendorf — APPROVED** [review]

**Comment by behlendorf** [comment]
@mcmilk good catch, I see things are passing in with Debian 11 in your QEMU PR against master so this does seem like a regression.  Pass CI run, https://github.com/openzfs/zfs/actions/runs/10899160824/job/30243940242

It does look we we can shed `config/kernel-get-user-pages.m4` but we can always follow up with a PR for this.

**Comment by robn** [comment]
Kernel for Debian 11 builders appears to be 5.10.223. That's in my test set, so it builds and passes basic sanity checks, so most likely I've removed something that caused it to use a non-working fallback or something like that. I'll check it out today or tomorrow.

**Comment by robn** [comment]
Last push should sort it.

The `iov_iter` tests are a little weird, they start out as "yes" and then test a few things, and if any fail, flips to "no". In the very first commit, I removed some "unused" tests, but that made a previously conditional test always get run, and then when it failed, disabled `HAVE_VFS_IOV_ITER`. So we didn't set iterator callbacks, and the kernel would fail various things.

Also added a commit removing `config/kernel-get-user-pages.m4` and unused defines, might as well while we're here.

**Comment by behlendorf** [comment]
Thanks for the quick fix there.  We did get a few failures but they're not new.

**Comment by behlendorf** [comment]
Merged thanks again for working through all of these config checks!

**Comment by robn** [comment]
Nice, thanks! And thanks everyone for the great feedback and testing. This will make things much easier going forward :)

**Comment by Harry-Chen** [comment]
As reported in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1098887, the `Linux-Minimum` is now 4.18. But there is [code using `SHIRNK_EMPTY`](https://github.com/openzfs/zfs/blob/7f05fface3a0a6790ca13d07743e7adc9e421469/module/os/linux/zfs/zfs_vfsops.c#L1218-L1242), which is [not in upstream Linux 4.18](https://elixir.bootlin.com/linux/v4.18.20/A/ident/SHRINK_EMPTY). Maybe RHEL / CentOS / Rocky have backported it, so zfs can compile on 4.18 kernels in these distros.

We may need to bump it to 4.19, or do some backport as well.
And how should we prevent such situation from happening again?

CC: @robn @behlendorf

**Comment by Fabian-Gruenbichler** [comment]
this contained a regression for zvol flush behavior, see https://github.com/openzfs/zfs/pull/17131
